File: UrnModel/AggrFrag.R
library(testit) # para pruebas, recuerda instalar antes de usar
k <- 10000
n <- 1000000
originales <- rnorm(k)
cumulos <- originales - min(originales) + 1
cumulos <- round(n * cumulos / sum(cumulos))
assert(min(cumulos) > 0)
diferencia <- n - sum(cumulos)
if (diferencia > 0) {
    for (i in 1:diferencia) {
        p <- sample(1:k, 1)
        cumulos[p] <- cumulos[p] + 1
    }
} else if (diferencia < 0) {
    for (i in 1:-diferencia) {
        p <- sample(1:k, 1)
        if (cumulos[p] > 1) {
            cumulos[p] <- cumulos[p] - 1
        }
    }
}
assert(length(cumulos[cumulos == 0]) == 0) # que no haya vacios
assert(sum(cumulos) == n)
c <- median(cumulos) # tamanio critico de cumulos
d <- sd(cumulos) / 4 # factor arbitrario para suavizar la curva
rotura <- function(x) {
    return (1 / (1 + exp((c - x) / d)))
}
union <- function(x) {
    return (exp(-x / c))
}
romperse <- function(tam, cuantos) {
    romper <- round(rotura(tam) * cuantos) # independientes
    resultado <- rep(tam, cuantos - romper) # los demas
    if (romper > 0) {
        for (cumulo in 1:romper) { # agregar las rotas
            t <- 1
            if (tam > 2) { # sample no jala con un solo valor
                t <- sample(1:(tam-1), 1)
            }
            resultado <- c(resultado, t, tam - t)
        }
    }
    assert(sum(resultado) == tam * cuantos) # no hubo perdidas
    return(resultado)
}
unirse <- function(tam, cuantos) {
    unir <- round(union(tam) * cuantos) # independientes
    if (unir > 0) {
        division <- c(rep(-tam, unir), rep(tam, cuantos - unir))
        assert(sum(abs(division)) == tam * cuantos)
        return(division)
    } else {
        return(rep(tam, cuantos))
    }
}
freq <- as.data.frame(table(cumulos))
names(freq) <- c("tam", "num")
freq$tam <- as.numeric(levels(freq$tam))[freq$tam]
duracion <- 50
digitos <- floor(log(duracion, 10)) + 1
for (paso in 1:duracion) {
    assert(sum(cumulos) == n)
    cumulos <- integer()
    for (i in 1:dim(freq)[1]) { # fase de rotura
        urna <- freq[i,]
        if (urna$tam > 1) { # no tiene caso romper si no se puede
            cumulos <- c(cumulos, romperse(urna$tam, urna$num))
        } else {
            cumulos <- c(cumulos, rep(1, urna$num))
        }
    }
    assert(sum(cumulos) == n)
    assert(length(cumulos[cumulos == 0]) == 0) # que no haya vacios
    freq <- as.data.frame(table(cumulos)) # actualizar urnas
    names(freq) <- c("tam", "num")
    freq$tam <- as.numeric(levels(freq$tam))[freq$tam]
    assert(sum(freq$num * freq$tam) == n)
    cumulos <- integer()
    for (i in 1:dim(freq)[1]) { # fase de union
        urna <- freq[i,]
        cumulos <- c(cumulos, unirse(urna$tam, urna$num))
    }
    assert(sum(abs(cumulos)) == n)
    assert(length(cumulos[cumulos == 0]) == 0) # que no haya vacios
    juntarse <- -cumulos[cumulos < 0]
    cumulos <- cumulos[cumulos > 0]
    assert(sum(cumulos) + sum(juntarse) == n)
    nt <- length(juntarse)
    if (nt > 0) {
        if (nt > 1) {
            juntarse <- sample(juntarse)
            for (i in 1:floor(nt / 2) ) {
                cumulos <- c(cumulos, juntarse[2*i-1] + juntarse[2*i])
            }
        }
        if (nt %% 2 == 1) {
            cumulos <- c(cumulos, juntarse[nt])
        }
    }
    assert(sum(cumulos) == n)
    freq <- as.data.frame(table(cumulos))
    names(freq) <- c("tam", "num")
    freq$tam <- as.numeric(levels(freq$tam))[freq$tam]
    assert(sum(freq$num * freq$tam) == n)
    tl <- paste(paso, "", sep="")
    while (nchar(tl) < digitos) {
        tl <- paste("0", tl, sep="")
    }
    png(paste("p8_ct", tl, ".png", sep=""), width=300, height=300)
    tope <- 50 * ceiling(max(cumulos) / 50)
    hist(cumulos, breaks=seq(0, tope, 50), 
         main=paste("Paso", paso, "con ambos fen\u{00f3}menos"), freq=FALSE,
         ylim=c(0, 0.05), xlab="Tama\u{00f1}o", ylab="Frecuencia relativa")
    graphics.off()
}

File: UrnModel/aggrFrag.py
# -*- coding: utf-8 -* # para incluir acentos
import numpy as np
from random import randint
 
k = 10000
n = 1000000
orig = np.random.normal(size = k)
cumulos = orig - min(orig)
cumulos += 1 # ahora el menor vale uno
cumulos = cumulos / sum(cumulos) # ahora suman a uno
cumulos *= n # ahora suman a n, pero son valores decimales
cumulos = np.round(cumulos).astype(int) # ahora son enteros
diferencia = n - sum(cumulos) # por cuanto le hemos fallado
cambio = 1 if diferencia > 0 else -1
while diferencia != 0:
    p = randint(0, k - 1)
    if cambio > 0 or (cambio < 0 and cumulos[p] > 0): # sin vaciar
        cumulos[p] += cambio
        diferencia -= cambio
assert all(cumulos != 0)
assert sum(cumulos) == n
 
c = np.median(cumulos) # tamaño crítico de cúmulos
d = np.std(cumulos) / 4 # factor arbitrario para suavizar la curva
 
from math import exp, floor, log
 
def rotura(x, c, d):
    return 1 / (1 + exp((c - x) / d))
 
def union(x, c):
    return exp(-x / c)
 
from random import random
 
def romperse(tam, cuantos):
    if tam == 1: # no se puede romper
        return [tam] * cuantos
    res = []
    for cumulo in range(cuantos):
        if random() < rotura(tam, c, d):
            primera = randint(1, tam - 1)
            segunda = tam - primera
            assert primera > 0
            assert segunda > 0
            assert primera + segunda == tam
            res += [primera, segunda]
        else:
            res.append(tam) # no rompió
    assert sum(res) == tam * cuantos
    return res
 
def unirse(tam, cuantos):
    res = []
    for cumulo in range(cuantos):
        if random() < union(tam, c):
            res.append(-tam) # marcamos con negativo los que quieren unirse
        else:
            res.append(tam)
    return res
 
from numpy.random import shuffle
import matplotlib.pyplot as plt
 
duracion = 50
digitos = floor(log(duracion, 10)) + 1
 
for paso in range(duracion):
    assert sum(cumulos) == n
    assert all([c > 0 for c in cumulos]) 
    (tams, freqs) = np.unique(cumulos, return_counts = True)
    cumulos = []
    assert len(tams) == len(freqs)
    for i in range(len(tams)):
        cumulos += romperse(tams[i], freqs[i]) 
    assert sum(cumulos) == n
    assert all([c > 0 for c in cumulos]) 
    (tams, freqs) = np.unique(cumulos, return_counts = True)
    cumulos = []
    assert len(tams) == len(freqs)
    for i in range(len(tams)):
        cumulos += unirse(tams[i], freqs[i])
    cumulos = np.asarray(cumulos)
    neg = cumulos < 0
    a = len(cumulos)
    juntarse = -1 * np.extract(neg, cumulos) # sacarlos y hacerlos positivos
    cumulos = np.extract(~neg, cumulos).tolist() # los demás van en una lista
    assert a == len(juntarse) + len(cumulos)
    nt = len(juntarse)
    if nt > 1:
        shuffle(juntarse) # orden aleatorio
    j = juntarse.tolist()
    while len(j) > 1: # agregamos los pares formados
        cumulos.append(j.pop(0) + j.pop(0))
    if len(j) > 0: # impar
        cumulos.append(j.pop(0)) # el ultimo no alcanzó pareja
    assert len(j) == 0
    assert sum(cumulos) == n
    assert all([c != 0 for c in cumulos])
    cortes = np.arange(min(cumulos), max(cumulos), 50)
    plt.hist(cumulos, bins = cortes, align = 'right', density = True)
    plt.xlabel('Tamaño')
    plt.ylabel('Frecuencia relativa')
    plt.ylim(0, 0.05)
    plt.title('Paso {:d} con ambos fenómenos'.format(paso + 1))
    plt.savefig('p8p_ct' + format(paso, '0{:d}'.format(digitos)) + '.png')
    plt.close()
